File: src/routes/spots.ts
import type { FastifyInstance } from "fastify";
import type { ZodTypeProvider } from "fastify-type-provider-zod";

import {
  AvailabilityResponseSchema,
  LotIdParamSchema,
} from "../schemas.js";
import { SpotType } from "../types/enums.js";
import type { AppContext } from "../app-context.js";

export async function spotsRoutes(fastify: FastifyInstance, ctx: AppContext) {
  const app = fastify.withTypeProvider<ZodTypeProvider>();

  app.get(
    "/lots/:lot_id/availability",
    {
      schema: {
        tags: ["availability"],
        summary: "Real-time per-spot-type availability for a lot.",
        params: LotIdParamSchema,
        response: { 200: AvailabilityResponseSchema },
      },
    },
    async (req) => {
      const counts = ctx.parking.availability(ctx.db, req.params.lot_id);
      const rows = (Object.values(SpotType) as SpotType[]).map((st) => ({
        spot_type: st,
        available: counts[st]?.available ?? 0,
        total: counts[st]?.total ?? 0,
      }));
      return { lot_id: req.params.lot_id, rows };
    },
  );
}

File: app/api/v1/spots.py
"""Real-time spot availability endpoints."""

from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.enums import SpotType
from app.schemas import AvailabilityResponse, AvailabilityRow
from app.services.parking_service import ParkingService

router = APIRouter(prefix="/lots/{lot_id}", tags=["availability"])


def get_service() -> ParkingService:
    return ParkingService()


@router.get("/availability", response_model=AvailabilityResponse)
def get_availability(
    lot_id: int,
    db: Session = Depends(get_db),
    service: ParkingService = Depends(get_service),
) -> AvailabilityResponse:
    counts = service.availability(db, lot_id=lot_id)
    rows = [
        AvailabilityRow(
            spot_type=spot_type,
            available=counts.get(spot_type, {}).get("available", 0),
            total=counts.get(spot_type, {}).get("total", 0),
        )
        for spot_type in SpotType
    ]
    return AvailabilityResponse(lot_id=lot_id, rows=rows)
